internal/mackerel/service.go

```diff
@@ -22,56 +22,53 @@ func ServiceNameValidator() validator.String {
 	)
 }
 
-type ServiceModel struct {
+type ServiceModel = ServiceModelV0
+
+type ServiceModelV0 struct {
 	ID   types.String `tfsdk:"id"`
-	Name types.String `tfsdk:"name"`
+	Name string       `tfsdk:"name"`
 	Memo types.String `tfsdk:"memo"`
 }
 
-func ReadService(ctx context.Context, client *Client, name string) (*ServiceModel, error) {
+func ImportService(_ context.Context, id string) (ServiceModel, error) {
+	return ServiceModelV0{
+		ID:   types.StringValue(id),
+		Name: id,
+	}, nil
+}
+
+func ReadService(ctx context.Context, client *Client, name string) (ServiceModel, error) {
 	return readServiceInner(ctx, client, name)
 }
 
 type serviceFinder interface {
 	FindServices() ([]*mackerel.Service, error)
 }
 
-func readServiceInner(_ context.Context, client serviceFinder, name string) (*ServiceModel, error) {
+func readServiceInner(_ context.Context, client serviceFinder, name string) (ServiceModel, error) {
```

Comment:
Why does this `readServiceInner` function take an unused `contextContext`.

Reply:
> Why does this readServiceInner function take an unused context.Context as an argument? If you don't need it, why not remove it?

Since this process involves network IO, it is preferable to receive `context`. Current Mackerel client does not handle `context` and `readServiceInner` does not use it, but there is no need to expose this to the outside.

> I have another but related question. Why is there a readServiceInner function separate from the ReadService function? func ReadService(_ context.Context, client *Client, name string) (ServiceModel, error) works, I think.

For testing, `readServiceInner` requires `serviceFinder` as a client. As `serviceFinder` is only for testing, the
service reading interface is splitted into internal (`readServiceInner`) and external one (`ReadService`).

Reply:
>For testing, readServiceInner requires serviceFinder as a client.

That's right. I missed the distinction between the `*Client` type and the `serviceFinder` type.

>Since this process involves network IO, it is preferable to receive context.

This is the reason that the `ReadService` public function receives the `context.Context`. `readServiceInner` function is private. It is natural for a public function to take a context (that context should be cancelable, not dummy) if it involves network IO, but I don't think a private function should take a fallback argument since it does not owe any promise to the outside world.

On a new point, is the `ReadService` function non-cancelable?

Reply:
You're right: private function signatures should represent implementation details (https://github.com/mackerelio-labs/terraform-provider-mackerel/pull/229/commits/b0dc3cf4a0468dee5d0c7de30ffedb9a4c81756c).
Also, currently public functions that calls API, such as `ReadService` are not cancelable and this should be documented (https://github.com/mackerelio-labs/terraform-provider-mackerel/pull/229/commits/8269bc2033d0f9e691426d12efaa8053b53e8337).